### ISSUE TITLE
How to log custom metric where compute returns a dict?

### ISSUE THREAD

**Issue by ChickenTarm** [PyTorchLightning/metrics#919]
## 🐛 Bug
I get this error when I try to log a metric whose return is of the form: {"metric_name_1": <metric value in the form of a torch.Tensor>, "metric_name_2": ...}

```

    Exception has occurred: AttributeError       (note: full exception trace is shown but execution is paused at: _run_module_as_main)
    'dict' object has no attribute 'detach'
      File "[/usr/local/lib/python3.9/dist-packages/pytorch_lightning/trainer/connectors/logger_connector/result.py]()", line 533, in _get_cache
        return cache.detach()
      File "[/usr/local/lib/python3.9/dist-packages/pytorch_lightning/utilities/apply_func.py]()", line 96, in apply_to_collection
        return function(data, *args, **kwargs)
      File "[/usr/local/lib/python3.9/dist-packages/pytorch_lightning/trainer/connectors/logger_connector/result.py]()", line 561, in metrics
        value = apply_to_collection(result_metric, ResultMetric, self._get_cache, on_step, include_none=False)
      File "[/usr/local/lib/python3.9/dist-packages/pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py]()", line 322, in metrics
        return self.trainer._results.metrics(on_step)
      File "[/usr/local/lib/python3.9/dist-packages/pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py]()", line 352, in progress_bar_metrics
        metrics = self.metrics["pbar"]
      File "[/usr/local/lib/python3.9/dist-packages/pytorch_lightning/trainer/trainer.py]()", line 2149, in progress_bar_metrics
        return self.logger_connector.progress_bar_metrics
      File "[/usr/local/lib/python3.9/dist-packages/pytorch_lightning/callbacks/progress/base.py]()", line 204, in get_metrics
        pbar_metrics = trainer.progress_bar_metrics
      File "[/usr/local/lib/python3.9/dist-packages/pytorch_lightning/callbacks/progress/tqdm_progress.py]()", line 242, in on_train_epoch_end
        self.main_progress_bar.set_postfix(self.get_metrics(trainer, pl_module))
      File "[/usr/local/lib/python3.9/dist-packages/pytorch_lightning/trainer/callback_hook.py]()", line 93, in on_train_epoch_end
        callback.on_train_epoch_end(self, self.lightning_module)
      File "[/usr/local/lib/python3.9/dist-packages/pytorch_lightning/trainer/trainer.py]()", line 1495, in call_hook
        callback_fx(*args, **kwargs)
      File "[/usr/local/lib/python3.9/dist-packages/pytorch_lightning/loops/epoch/training_epoch_loop.py]()", line 298, in on_run_end
        self.trainer.call_hook("on_train_epoch_end")
      File "[/usr/local/lib/python3.9/dist-packages/pytorch_lightning/loops/base.py]()", line 151, in run
        output = self.on_run_end()
      File "[/usr/local/lib/python3.9/dist-packages/pytorch_lightning/loops/fit_loop.py]()", line 234, in advance
        self.epoch_loop.run(data_fetcher)
      File "[/usr/local/lib/python3.9/dist-packages/pytorch_lightning/loops/base.py]()", line 145, in run
        self.advance(*args, **kwargs)
      File "[/usr/local/lib/python3.9/dist-packages/pytorch_lightning/trainer/trainer.py]()", line 1319, in _run_train
        self.fit_loop.run()
      File "[/usr/local/lib/python3.9/dist-packages/pytorch_lightning/trainer/trainer.py]()", line 1289, in run_stage
        return self._run_train()
      File "[/usr/local/lib/python3.9/dist-packages/pytorch_lightning/plugins/training_type/training_type_plugin.py]()", line 202, in start_training
        self._results = trainer.run_stage()
      File "[/usr/local/lib/python3.9/dist-packages/pytorch_lightning/trainer/trainer.py]()", line 1279, in _dispatch
        self.training_type_plugin.start_training(self)
      File "[/usr/local/lib/python3.9/dist-packages/pytorch_lightning/trainer/trainer.py]()", line 1199, in _run
        self._dispatch()
      File "[/usr/local/lib/python3.9/dist-packages/pytorch_lightning/trainer/trainer.py]()", line 777, in _fit_impl
        self._run(model, ckpt_path=ckpt_path)
      File "[/usr/local/lib/python3.9/dist-packages/pytorch_lightning/trainer/trainer.py]()", line 685, in _call_and_handle_interrupt
        return trainer_fn(*args, **kwargs)
      File "[/usr/local/lib/python3.9/dist-packages/pytorch_lightning/trainer/trainer.py]()", line 740, in fit
        self._call_and_handle_interrupt(
      File "[/home/experimentation/ml_system/train.py]()", line 138, in main
        trainer.fit(
      File "[/usr/local/lib/python3.9/dist-packages/clearml/binding/hydra_bind.py]()", line 146, in _patched_task_function
        return task_function(a_config, *a_args, **a_kwargs)
      File "[/usr/local/lib/python3.9/dist-packages/hydra/core/utils.py]()", line 160, in run_job
        ret.return_value = task_function(task_cfg)
      File "[/usr/local/lib/python3.9/dist-packages/hydra/core/utils.py]()", line 233, in return_value
        raise self._return_value
      File "[/usr/local/lib/python3.9/dist-packages/hydra/_internal/hydra.py]()", line 111, in run
        _ = ret.return_value
      File "[/usr/local/lib/python3.9/dist-packages/clearml/binding/hydra_bind.py]()", line 89, in _patched_hydra_run
        return PatchHydra._original_hydra_run(self, config_name, task_function, overrides, *args, **kwargs)
      File "[/usr/local/lib/python3.9/dist-packages/hydra/_internal/utils.py]()", line 378, in <lambda>
        lambda: hydra.run(
      File "[/usr/local/lib/python3.9/dist-packages/hydra/_internal/utils.py]()", line 214, in run_and_report
        raise ex
      File "[/usr/local/lib/python3.9/dist-packages/hydra/_internal/utils.py]()", line 214, in run_and_report
        raise ex
      File "[/usr/local/lib/python3.9/dist-packages/hydra/_internal/utils.py]()", line 377, in _run_hydra
        run_and_report(
      File "[/usr/local/lib/python3.9/dist-packages/hydra/main.py]()", line 48, in decorated_main
        _run_hydra(
      File "[/home/experimentation/ml_system/train.py]()", line 147, in <module>
        main()
      File "[/usr/lib/python3.9/runpy.py]()", line 87, in _run_code
        exec(code, run_globals)
      File "[/usr/lib/python3.9/runpy.py]()", line 97, in _run_module_code
        _run_code(code, mod_globals, init_globals,
      File "[/usr/lib/python3.9/runpy.py]()", line 268, in run_path
        return _run_module_code(code, init_globals, run_name,
      File "[/usr/lib/python3.9/runpy.py]()", line 87, in _run_code
        exec(code, run_globals)
      File "[/usr/lib/python3.9/runpy.py]()", line 197, in _run_module_as_main (Current frame)
        return _run_code(code, main_globals, None
```

### Expected behavior

If a metric class returns multiple metrics like the detection metric that is built-in, the logging should iterate over the returned metrics and log them individually.

### Environment

CUDA:
- GPU:
- Tesla T4
- available: True
- version: 11.4
Packages:
- numpy: 1.21.4
- PyTorch_version: 1.11.0+cu113
- pytorch-lightning: 1.5.10
System:
- OS: Linux
- architecture:
- 64bit
- ELF
- processor: x86_64
- python: 3.9.7

**Comment by SkafteNicki** [comment]
Hi @ChickenTarm ,
If what you are trying to log is a dict where each value is a *scalar* tensor, then you simply need to replace `self.log` with
```python
self.log_dict(metric_dict)
```
where each metric will then be logged under its respective key in the dict. If some of the values are not scalar tensors you will need to pop those from the dict and then manually log them.
